linenoise.c

```diff
@@ -794,10 +822,44 @@ static int linenoiseEdit(int stdin_fd, int stdout_fd, char *buf, size_t buflen,
     if (write(l.ofd,prompt,l.plen) == -1) return -1;
     while(1) {
         char c;
-        int nread;
+        int nread = 0;
         char seq[3];
 
-        nread = read(l.ifd,&c,1);
+        fd_set rfds;
+        int retval;
+
+        FD_ZERO(&rfds);
+        FD_SET(l.ifd, &rfds);
+        FD_SET(ctrl_pipe_read_fd, &rfds);
+        int nfds = (l.ifd > ctrl_pipe_read_fd) ? l.ifd+1 : ctrl_pipe_read_fd+1; //compute 1+(the greater fd) for select
+
+        retval = select(nfds, &rfds, NULL, NULL, NULL);
+        if (retval == -1)
+        {
+            fprintf(stderr, "Problem with select in linenoise\n");
+            continue;
+        }
+        else if (retval)
+        {
+            if (FD_ISSET(l.ifd, &rfds))
+            {
+                nread = read(l.ifd,&c,1);
+            }
+            else if (FD_ISSET(ctrl_pipe_read_fd, &rfds))
+            {
+                nread = read(ctrl_pipe_read_fd,&c,1);
+            }
+            else
+            {
+                continue;
```

Comment:
When does this happen? Since you only pass readfds wouldn't retval being non-zero indicate one of them is set? Should this be a FATAL_ERROR? At least a comment?